src/components/controls/controls.js

```diff
@@ -0,0 +1,220 @@
+    this.yaw = new THREE.Object3D();
+    this.yaw.position.y = 10;
+    this.yaw.add(this.pitch);
+    this.heading = new THREE.Euler(0, 0, 0, 'YXZ');
+
+    if (this.el.sceneEl.hasLoaded) {
+      this.validateControls();
+    } else {
+      this.el.sceneEl.addEventListener('loaded', this.validateControls.bind(this));
+    }
+  },
+
+  update: function () {
+    // Re-validate if scene has loaded. If not, listener was already bound in init().
+    if (this.el.sceneEl.hasLoaded) {
+      this.validateControls();
+    }
+  },
+
+  tick: function (t, dt) {
+    if (isNaN(dt)) { return; }
+
+    // Update rotation.
+    if (this.data.rotationEnabled) {
+      this.updateRotation(dt);
+    }
+
+    // Update velocity. If FPS is too low, reset.
+    if (this.data.positionEnabled && dt / 1000 > MAX_DELTA) {
+      this.velocity.set(0, 0, 0);
+      this.el.setAttribute('velocity', this.velocity);
+    } else if (this.data.positionEnabled) {
+      this.updateVelocity(dt);
+    }
+  },
+
+  /**
+   * Updates rotation based on input from the active rotation component, if any.
+   * @param  {number} dt
+   */
+  updateRotation: function (dt) {
+    var control, dRotation;
+
+    control = this.getActiveRotationControls();
+    if (control && control.getRotationDelta) {
+      dRotation = control.getRotationDelta(dt);
+      this.yaw.rotation.y -= dRotation.x;
+      this.pitch.rotation.x -= dRotation.y;
+      this.pitch.rotation.x = Math.max(-PI_2, Math.min(PI_2, this.pitch.rotation.x));
+      this.el.setAttribute('rotation', {
+        x: THREE.Math.radToDeg(this.pitch.rotation.x),
+        y: THREE.Math.radToDeg(this.yaw.rotation.y),
+        z: 0
+      });
+    } else if (control) {
+      this.el.setAttribute('rotation', control.getRotation());
+    }
+  },
+
+  /**
+   * Updates velocity based on input from the active position component, if any.
+   * @param {number} dt
+   */
+  updateVelocity: function (dt) {
+    var control;
+    var data = this.data;
+
+    if (this.el.hasAttribute('velocity')) {
+      this.velocity.copy(this.el.getComputedAttribute('velocity'));
+    }
+
+    this.velocity.x -= this.velocity.x * data.positionEasing * dt / 1000;
+    this.velocity.z -= this.velocity.z * data.positionEasing * dt / 1000;
+
+    control = this.getActivePositionControls();
+    if (control && control.getVelocityDelta) {
+      this.applyVelocityDelta(dt, control.getVelocityDelta(dt));
+    } else if (control) {
+      this.velocity.copy(control.getVelocity());
+    }
+
+    this.el.setAttribute('velocity', {x: this.velocity.x, y: this.velocity.y, z: this.velocity.z});
+  },
+
+  /**
+   * Updates internal velocity. Takes a velocity delta (relative to current heading), rotates it
+   * to the heading, and calculates velocity in world space.
+   * @param {number} dt
+   * @param {THREE.Vector3} dVelocity
+   */
+  applyVelocityDelta: function (dt, dVelocity) {
+    var data = this.data;
+    var rotation = this.el.getAttribute('rotation');
```

Comment:
It's starting to feel more intuitive to me that `getAttribute` should return all data since that is the most common case, and then we create a special method for `getDefinedAttribute` or something. Not sure the `computedStyle` analog is worth it.